### ISSUE TITLE
Clipping a Mesh by a Plane?

### ISSUE THREAD

**Issue by RaphiePS** [nicklockwood/Euclid#6]
Heyo, thanks for the incredible library! I'd like to clip a `Mesh` by a `Plane`, exactly like this [CGAL function](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__corefinement__grp.html#ga5bf228a516a46f6d1c80937067a17f7e) explained in more detail [here](https://doc.cgal.org/latest/Polygon_mesh_processing/index.html#coref_clip).

I see that `Polygon` has both `split` and `clip` methods, but it's unclear exactly how I'd use these to clip an entire `Mesh`. Is it possible to compose these methods into a whole-mesh clipper, or am I better off implementing mesh clipping from the ground up? Thanks again!

**Comment by nicklockwood** [comment]
@RaphiePS you can clip a mesh by creating a cube with one side matching the plane you want to clip along and then subtracting the cube from your mesh using the CSG functions.

It would be possible to clip directly to a plane using private methods. I can look at exposing a plane clipping method in the public API if the cube solution is unsatisfactory.

**Comment by RaphiePS** [comment]
@nicklockwood thanks for responding so quickly - I'll give it a try! The cube approach is really clever and I'll give it a shot. I assume there's no performance penalty for using a _massive_ cube to simulate an infinite plane, but I'll report back. 

Cube-clipping should definitely work, though it does seem just a tad hacky. If it's not too much effort a public plane-clip method would be awesome. If it's a pain in the butt, don't bother 😄

**Comment by nicklockwood** [comment]
@RaphiePS the size of the cube should make no difference - bigger is actually better as clipping will be more efficient if the 5 other faces of the cube are far away from the mesh being clipped.

I've just realized though that this approach will fill in the clipped surface of the clipped mesh with additional polygons (rather than leaving the backfaces exposed), which may not be what you want.

**Comment by andygeers** [comment]
I too would love a method for clipping by a plane if it was easy to expose those private methods, Nick! Otherwise I might just submit a PR that makes the cube method available

**Comment by nicklockwood** [comment]
@andygeers it should be pretty simple, I'll try to put up an implementation asap.

**Comment by andygeers** [comment]
Thanks Nick, that would be great. Filling in the clipped surface is a good thing in my use case.

**Comment by nicklockwood** [comment]
@andygeers I've added a method to clip without filling the clipped surface (it's on the develop branch). Filling is slightly more complicated but should also be doable.

Using the cube solution in the meantime should also work fine. It's probably not worth adding a PR for it because it's not really the optimal solution, and the hard part (automatically detecting the right size to make the cube) is also the hard part when it comes to doing it correctly.

**Comment by andygeers** [comment]
Thanks Nick. Well if I figure out that hard part for the cube method I'll at least send you a link to my code in case it's of benefit. Once we get a way to fill in the hole, I can also use it to implement the other bit I've been working on - the *intersection* between a mesh and a plane, which returns an array of polygons. I can just clip by the plane then return all polygons contained in the plane.

**Comment by nicklockwood** [comment]
@andygeers when you say "intersection between a mesh and a plane", do you mean the polygons from the original mesh that intersect the plane, or the polygons that make up the area on the surface of the plane that is inside the mesh?

The former can be done quite easily, although it currently requires private methods. I just added an internal ` compare(with plane: Plane)` method to Polygon which will tell you if it intersects.

The latter can already be done with the public API by creating a large double-sided rectangle (i.e. a cube with zero depth) and taking the union of that and your other mesh.

**Comment by andygeers** [comment]
Ooh, good tip there - I meant the latter, and hadn't considered that approach. Nice!

**Comment by andygeers** [comment]
(can that be used as a solution to your other problem of how to fill in the hole after clipping by the plane?)

**Comment by nicklockwood** [comment]
@andygeers that would certainly form part of the solution, but that's why we need to know the size of the cube.

It's honestly not a very hard problem, I just haven't thought about it yet - it probably involves something like taking the bounding box of the mesh and then rotating it to match the coordinate space of the plane.

**Comment by andygeers** [comment]
I haven't got this fully working, but here's my starter for ten on an approach for determining the cube size and position:

```
private func distanceBetween(point: Vector, plane: Plane) -> Double {
        return (point - plane.pointOnPlane).dot(plane.normal)
    }
    
    private func determineClippingCubeSize(plane: Plane) -> Double {
        // Project each corner of our bounds on to the plane along its normal
        let distanceMin = distanceBetween(point: self.bounds.min, plane: plane)
        let distanceMax = distanceBetween(point: self.bounds.max, plane: plane)
        let minOnPlane = self.bounds.min - plane.normal * distanceMin
        let maxOnPlane = self.bounds.max - plane.normal * distanceMax
        
        // Now return the distance between them
        return (maxOnPlane - minOnPlane).length
    }
    
    private func clippingCubePosition(size: Double, plane: Plane) -> Vector {
        // Project centre on to plane
        let center = self.bounds.center
        let distanceToPlane = distanceBetween(point: center, plane: plane)
        let centerOnPlane = center - plane.normal * distanceToPlane
        
        // Now move away so that that point is on the edge of the cube
        return centerOnPlane - plane.normal * (size / 2.0)
    }
```

**Comment by andygeers** [comment]
Ah, so I was almost right: cube position should be `return centerOnPlane + plane.normal * (size / 2.0)` instead. Admittedly have only tested so far with meshes that are axis aligned to start with, but seems to work nicely

**Comment by nicklockwood** [comment]
@andygeers that didn't quite work as the min/max of the bounds aren't necessarily the maximum distance for all plane angles. I've managed to find a solution though, and I've pushed it to the develop branch.

To try it out, in the Example app use something like this:

```swift
// create some geometry using Euclid
let sphere = Mesh.sphere(slices: 40, material: UIColor.blue)
let plane = Plane(normal: Vector(0, 1, 0), pointOnPlane: .zero)!
let mesh = sphere.clip(to: plane, fill: UIColor.green)
```

**Comment by nicklockwood** [comment]
Implemented in 0.2.1
